crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -69,6 +72,104 @@ struct CheckpointStoreTables {
 
     /// Stores certified checkpoints
     certified_checkpoints: DBMap<CheckpointSequenceNumber, CertifiedCheckpointSummary>,
+    /// Map from checkpoint digest to certified checkpoint
+    checkpoint_by_digest: DBMap<CheckpointDigest, CertifiedCheckpointSummary>,
+
+    /// Watermarks used to determine the highest verified and fully synced checkpoints
+    watermarks: DBMap<CheckpointWatermark, (CheckpointSequenceNumber, CheckpointDigest)>,
+}
+
+impl CheckpointStoreTables {
+    pub fn get_checkpoint_by_digest(
+        &self,
+        digest: &CheckpointDigest,
+    ) -> Option<VerifiedCheckpoint> {
+        self.checkpoint_by_digest
+            .get(digest)
+            .ok()
+            .flatten()
+            .map(VerifiedCheckpoint::new_unchecked)
+    }
+
+    pub fn get_checkpoint_by_sequence_number(
+        &self,
+        sequence_number: CheckpointSequenceNumber,
+    ) -> Option<VerifiedCheckpoint> {
+        self.certified_checkpoints
+            .get(&sequence_number)
+            .ok()
```

Comment:
Here and in other places in this implementation we seem to ignore storage error and treat it as None.

This does not sounds like reasonable behavior to me. I think options are
(a) Panic on storage error(@lxfind probably have opinion on that option)
(b) At least log error, and then return None
(c) (My favorite) Change interface so all those places return SuiError and propagate it to the caller.

Reply:
Yeah i figured this would be a point to discuss. I guess the question is: if a read/write to the db fails what do we actually want to do? I do agree that returning the Error is probably the better thing to do, it does just push the "what do we do on error" up the stack. For example whats the expected behavior if we fail to insert a new checkpoint into the db?

Reply:
> if a read/write to the db fails what do we actually want to do?

I think there are different opinions here.
I think it is ok to panic on storage error, but I last time we chatted with @lxfind about it he preferred not to panic when possible, in which case we can either retry (for tasks like checkpoint builder/aggregator) or return to user for RPC handlers.

Reply:
Does anyone know what are the possible conditions when db read/write could fail? And whether they will tend to be permanent or intermittent? If they are intermittent then we should certainly consider having retries.
I guess in the end there will be a few critical locations where, if a db failure happens and if the condition is permanent, we must crash because there is no better thing to do.

Reply:
Perhaps we can resolve it as following:

I personally believe that most if not all db read/write failures are pretty much non-retriable in practice, as such I think it is ok to just commonly panic on storage error.

If at some point we discover that there is an intermittent storage error that needs to be retried, DBMap will be taking care of that retry

Reply:
I just pushed a commit on top of this PR that does make these operations failable and pushes the panics up the call stack. Although given some of the developments on this conversation we may want to do something else now?